src/render/src/macros/mod.rs

```diff
@@ -31,7 +31,107 @@ macro_rules! gfx_format {
+///         out_color: gfx::RenderTarget<gfx::format::Rgba8> = "Target0",
+///         out_depth: gfx::DepthTarget<gfx::format::DepthStencil> = 
+///             gfx::preset::depth::LESS_EQUAL_WRITE,
+///     }
+/// }
+///
+/// impl Vertex {
+///     fn new(p: [i8; 3], tex: [i8; 2]) -> Vertex {
+///         Vertex {
+///             pos: [p[0] as f32, p[1] as f32, p[2] as f32, 1.0f32],
+///             tex_coord: [tex[0] as f32, tex[1] as f32],
+///         }
+///     }
+/// }
+///
+/// fn main() {
+///     let vertex_data = [
+///         Vertex::new([-1, -1, 1], [0, 0]),
+///         Vertex::new([ 1, -1, 1], [1, 0]),
+///         // Add more vertices..
+///     ];
+/// }
+/// ```
+/// `vertex` and `constant` structures defined with `gfx_defines!`
+/// can be extended with attributes:
+///
+/// ```{.rust}
+/// #[macro_use]
+/// extern crate gfx;
+///
+/// gfx_defines! {
+///     #[derive(Default)]
+///     vertex Vertex {
+///         pos: [f32; 4] = "a_Pos",
+///         tex_coord: [f32; 2] = "a_TexCoord",
+///     }
+/// }
+///
+/// fn main() {
+///     let vertex = Vertex::default();
+///     assert_eq!(vertex.pos[0], 0f32);
+///     assert_eq!(vertex.tex_coord[0], 0f32);
+/// }
+/// ```
+///
+/// # `pipe`
+///
+/// The `pipeline state object` or `pso` can consist of the following
+/// `pso` components:
+///
+/// - A single [vertex buffer](pso/buffer/type.VertexBuffer.html) component to hold the vertices.
+/// - A single [instance buffer](pso/buffer/type.InstanceBuffer.html) component.
+/// - Single or multiple [constant buffer](pso/buffer/struct.ConstantBuffer.html) components. (DX11 and OpenGL3)
+/// - Single or multiple [global buffer](pos/buffer/struct.Global.html) components.
+/// - Single or multiple [samplers](pos/resource/struct.Sampler.html) 
+///   such as [texture samplers](pso/resource/struct.TextureSampler.html].
+/// - [Render](pso/target/struct.RenderTarget.html), [blend](pso/target/struct.BlendTarget.html), 
+///   [depth](pso/target/struct.DepthTarget.html) and [stencil](pso/target/struct.StencilTarget.html) targets
+/// - A [shader resource view](pso/resource/struct.ShaderResource.html) (SRV, DX11)
+/// - A [unordered access view](pso/resource/struct.UnorderedAccess.html) (UAV, DX11, not yet implemented in the OpenGL backend)
+/// - A [scissor](pso/target/struct.Scissor.html) rectangle value (DX11)
+///
+/// Structure of a `pipeline state object` can be defined
+/// freely but should at the very least consist of one vertex buffer object.
+///
+/// # `vertex`
+///
+/// Defines a vertex format to be passed onto a vertex buffer. Similar
+/// to `pipeline state objects` multiple vertex formats can be set.
+///
+/// # `constant`
+///
+/// Defines a structure for shader constant data. This constant data
+/// is then appended into a constant buffer in the `pso`. Constant buffers
+/// are only supported by DirectX 11 and OpenGL backend uses it's own
+/// Uniform Buffer Object which requires OpenGL 3.1.
```

Comment:
It sounds like GL doesn't support the constant buffers. It does, just names them as UBOs. You may just say that constant buffers are supported on D3D11 and GL3.